### PR TITLE
feat: added graceful shutdown (#317) 

### PR DESCRIPTION
**Pull request by inigo-garcia**
Issue link: https://github.com/fastify/fastify-cli/issues/317

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test/start.test.js [hunk 1]
```diff
@@ -9,6 +9,7 @@ const semver = require('semver')
 const baseFilename = path.join(__dirname, 'fixtures', `test_${crypto.randomBytes(16).toString('hex')}`)
 const { fork } = require('child_process')
 const moduleSupport = semver.satisfies(process.version, '>= 14 || >= 12.17.0 < 13.0.0')
+require('events').EventEmitter.defaultMaxListeners = 0
```
Comment (kibertoad) [inline test/start.test.js:12]: can you add comment to this line because it's not self-explanatory why it's there?

Comment (inigo-garcia) [reply test/start.test.js:12]: I was running into these warnings when I was testing the implementation. `MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 11 SIGTERM listeners added to [process]. Use emitter.setMaxListeners() to increase limit`

I tried with: afterEachTest

Comment (kibertoad) [reply test/start.test.js:12]: If we are actually leaking listeners somewhere, that's rather bad, we are likely to cause warnings for user tests downstream then.

File: templates/eject-ts/server.ts [hunk 2]
```diff
@@ -7,6 +7,9 @@ dotenv.config();
 // Require the framework
 import Fastify from "fastify";
 
+// Require libraty to exit fastify process, gracefully (if possible)
```
Comment (simoneb) [inline templates/eject-ts/server.ts:10]:
```suggestion
// Require library to exit fastify process, gracefully (if possible)
```

File: templates/eject/server.js [hunk 3]
```diff
@@ -6,6 +6,9 @@ require('dotenv').config()
 // Require the framework
 const Fastify = require('fastify')
 
+// Require libraty to exit fastify process, gracefully (if possible)
```
Comment (simoneb) [inline templates/eject/server.js:9]:
```suggestion
// Require library to exit fastify process, gracefully (if possible)
```

**Review by simoneb — APPROVED** [review]
code wise lgtm, does it work? :)

File: templates/eject/server.js (hunk 3 above)
Comment (inigo-garcia) [reply templates/eject/server.js:9]: Thanks! I didn't notice it

**Comment by inigo-garcia** [comment]
i tried sending SIGTERM and SIGINT signals and it handles them correctly.

**Comment by simoneb** [comment]
@inigo-garcia please link the issue in the PR description

**Review by mcollina — CHANGES_REQUESTED** [review]
This needs tests

File: test/start.test.js (hunk 1 above)
Comment (mcollina) [reply test/start.test.js:12]: Unfortunately this means that we are not cleaning things up properly. You might need to call `uninstall` - see https://github.com/mcollina/close-with-grace#return-values.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: test/start.test.js (hunk 1 above)
Comment (inigo-garcia) [reply test/start.test.js]: Thanks for the feedback, I've included the `uninstall` method inside of `onClose` hook and we don't get any memory leak warnings

File: start.js [hunk 4]
```diff
@@ -126,6 +127,18 @@ async function runFastify (args) {
...
+  const closeListeners = closeWithGrace({ delay: 500 }, async function ({ signal, err, manual }) {
+    if (err) {
+      fastify.log.error(err)
+    }
+    await fastify.close()
+  })
+
+  await fastify.addHook('onClose', async (instance, done) => {
+    closeListeners.uninstall()
+    done()
```
Comment (mcollina) [inline start.js:139]: Don't call `done()` when using `async`.

**Review by mcollina — CHANGES_REQUESTED** [review]

File: start.js (hunk 4 above)
Comment (inigo-garcia) [reply start.js:139]: Thanks @mcollina , I'll update it

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by inigo-garcia** [comment]
@mcollina  @kibertoad @simoneb Thanks for your feedback

**Review by kibertoad — APPROVED** [review]

File: templates/eject/server.js [hunk 5]
```diff
@@ -15,6 +18,19 @@ const app = Fastify({
...
 
+// delay is the number of milliseconds for the graceful close to finish
+const closeListeners = closeWithGrace({ delay: 500 }, async function ({ signal, err, manual }) {
+  if (err) {
+    app.log.error(err)
+  }
+  await app.close()
+})
+
+app.addHook('onClose', async (instance, done) => {
```
[the remark this answers is not in the document]
Comment (simoneb) [reply templates/eject/server.js:29]: no special reason I suppose

**Pull request merged**